### PR TITLE
feat: add open_as expression

### PR DESCRIPTION
**Pull request by mlutze**
related to #5229

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Can we merge `Open` and `OpenAs`?

File: main/src/ca/uwaterloo/flix/language/phase/inference/RestrictableChooseInference.scala [hunk 1]
```diff
@@ -318,6 +318,41 @@ object RestrictableChooseInference {
       } yield (constrs, tpe, pur, eff)
   }
 
+  /**
+    * Performs type inference on the given OpenAs expression
+    */
+  def inferOpenAs(exp0: KindedAst.Expression.OpenAs, root: KindedAst.Root)(implicit flix: Flix): InferMonad[(List[Ast.TypeConstraint], Type, Type, Type)] = exp0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/inference/RestrictableChooseInference.scala:324]: I think we need some more documentation on all the return types.

**Comment by JonathanStarup** [comment]
> Can we merge `Open` and `OpenAs`?

I think we need to remove the choose* body check before we remove the old open

**Comment by mlutze** [comment]
> Can we merge `Open` and `OpenAs`?

It's a bit annoying to write `open_as Red Red` though

**Comment by mlutze** [comment]
> > Can we merge `Open` and `OpenAs`?
> 
> It's a bit annoying to write `open_as Red Red` though

Maybe the syntax could be `open X [as Y]` where the Y part is required if X is not a tag

**Comment by magnus-madsen** [comment]
> > > Can we merge `Open` and `OpenAs`?
> > 
> > 
> > It's a bit annoying to write `open_as Red Red` though
> 
> Maybe the syntax could be `open X [as Y]` where the Y part is required if X is not a tag

Lets do that and then I can merge.

**Comment by JonathanStarup** [comment]
> > Can we merge `Open` and `OpenAs`?
> 
> It's a bit annoying to write `open_as Red Red` though

open_as Color Red

**Comment by mlutze** [comment]
> > > Can we merge `Open` and `OpenAs`?
> > 
> > 
> > It's a bit annoying to write `open_as Red Red` though
> 
> `open_as Color Red`?

Good point. Still annoying though.

**Comment by mlutze** [comment]
It looks like we need both of them. I tried removing `open` and replacing it with the `open_as` construct. This is the result:

```
-- Type Error -------------------------------------------------- main/test/flix/Test.Exp.Choose.Recursive.flix

>> Unable to unify the Boolean formulas: '{Cst} + c0' and '{Not}'.

67 |         let cstOrNot = if (true) open Expr.Cst(true) else open Expr.Not(open Expr.Cst(true));
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                                    mismatched boolean formulas.

Type One: {Cst} + c0
Type Two: {Not}
```

The issue is that previously, `open` applied specifically to the `Expr.Not` tag, opening its index both for input and output. With `open_as`, only the output is opened.

**Pull request merged**